Fix issues

dientc opened the pull request:
Create a pull request for fixing issues #
- Implement cDriveManager as a singleton
- Remove device handle reference after call cDrive->IssueCommand() by let device handle as a parameter in CommandHandler and Protocol  IssueCommand()
- Avoid duplicating a drive when enumerating use a common device paths.
  Tested with some Ata commands

dientc committed:
Generated enum ScanForHardwareChanges and used BuildEventer tool

dientc committed:
Merge pull request #2 from virtium/master

Update changes from vtStor Virtium.

dientc committed:
Merge pull request #3 from virtium/master

Sync

dientc committed:
Fix issues:
+ Implement cDriveManager as a singleton
+ Remove device handle reference after call cDrive->IssueCommand() by let device handle as a parameter in CommandHandler and Protocol  IssueCommand()
+ Avoid duplicating a drive when enumerating use a common device paths.
Tested with some Ata commands

phandinhlan commented on vtStor.ATest/vtStor.ATest.vcxproj:
```diff
@@ -72,22 +72,22 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">
     <LinkIncremental>true</LinkIncremental>
-    <OutDir>$(ProjectDir)$(Platform)$(Configuration)\</OutDir>
+    <OutDir>$(SolutionDir)$(Platform)$(Configuration)\</OutDir>
```
Why are these changes to SolutionDir?

dientc replied:
I change it to test directly, if not changing, it needs to copy the output file into the destination to run. I discussed on it with anh Tam.

phandinhlan commented on vtStor/DriveManager.cpp:
```diff
@@ -17,8 +17,12 @@ limitations under the License.
+
 namespace vtStor
 {
+    cDriveManager* cDriveManager::single = nullptr;
```
This is usually named "instance".

phandinhlan commented on vtStor/DriveManager.cpp:
```diff
@@ -26,10 +30,21 @@ cDriveManager::cDriveManager()
     m_Drives.clear();
 }
 
+std::unique_ptr<cDriveManager> cDriveManager::getInstance()
```
Why is naming convention for function changed?

phandinhlan replied:
Have you actually try to call this function more than once?

phandinhlan commented on vtStor/DriveManager.cpp:
```diff
@@ -26,10 +30,21 @@ cDriveManager::cDriveManager()
 
+std::unique_ptr<cDriveManager> cDriveManager::getInstance()
+{
+    if (!single)
```
Stay with coding convention: if ( nullptr != single )

phandinhlan commented on vtStor/DriveManager.cpp:
```diff
@@ -26,10 +30,21 @@ cDriveManager::cDriveManager()
 
 cDriveManager::~cDriveManager()
 {
+    single = nullptr;
```
This is wrong. There needs to be a different way to "delete" the singleton.

phandinhlan commented on vtStor/DriveManager.h:
```diff
@@ -35,8 +35,14 @@ class cDriveManager : public cDriveManagerInterface
+    static cDriveManager* single;
+
+public:
+    static std::unique_ptr<cDriveManager> getInstance();
```
Put static method at the top.

phandinhlan commented on vtStorAta/DriveEnumeratorAta.cpp:
```diff
@@ -32,16 +32,13 @@ cDriveEnumeratorAta::~cDriveEnumeratorAta()
 }
 
-eErrorCode cDriveEnumeratorAta::EnumerateDrives( Vector_Drives& AddToList, U32& Count )
+eErrorCode cDriveEnumeratorAta::EnumerateDrives(std::vector<String> PathsList, Vector_Drives& AddToList, U32& Count )
```
This is passed by value, that means a new vector is create and everything copied over.

dientc committed:
Fix coding convention errors
Use pass by reference for m_DevicePaths

tamvirtium commented on vtStor/DriveManager.cpp:
```diff
@@ -17,8 +17,12 @@ limitations under the License.
+
 namespace vtStor
 {
+    cDriveManager* cDriveManager::m_sInstance = nullptr;
```
Why don't use smart pointer to store instance instead so that we don't have to care about how to free it?

phandinhlan commented on vtStorAta/DriveEnumeratorAta.cpp:
```diff
@@ -64,6 +61,8 @@ eErrorCode cDriveEnumeratorAta::EnumerateDrives( Vector_Drives& AddToList, U32&
             std::shared_ptr<cDriveInterface> drive = std::make_shared<cDriveAta>(std::make_shared<String>(devicePath));
 
             AddToList.push_back( drive );
+            devicePath.empty();
```
How does this prevent the next enumerator from trying to use it? You're relying on it to fail?

dientc replied:
Yes, I hope it to be fail when checking.
errorCode = GetStorageDeviceHandle( devicePath, deviceHandle );
        if ( eErrorCode::None != errorCode )
        {
            //TODO: handle error
            continue;
        }

phandinhlan commented on vtStorAtaProtocol/Platform/Windows/ProtocolAtaPassThrough.cpp:
```diff
@@ -50,7 +50,7 @@ namespace Protocol
                 cEssenseAta1 essense = cEssenseAta1::Reader(Essense);
                 
-                m_DeviceHandle = essense.GetDeviceHandle();
+                m_DeviceHandle = Handle;
```
Does it need to store the device handle as a member? What if the IssuePassThroughDirectCommand method takes in the handle as parameter?

dientc committed:
- Add ResetInstance() to destroy a singleton
- CloseHandle() when remove a drive (not yet for multi-platform)

dientc committed:
- Change the way to enumerate drives for using Range-based for loop
- Remove m_deviceHandle from Protocol class, we pass device handle as a parameter.

phandinhlan commented on vtStor/Drive.cpp:
```diff
@@ -31,6 +31,11 @@ cDrive::cDrive(std::shared_ptr<String> DevicePath) : m_DevicePath(DevicePath)
 
 cDrive::~cDrive()
 {
+    if (m_DeviceHandle)
```
This is wrong.

phandinhlan commented on vtStor/DriveEnumeratorInterface.h:
```diff
@@ -33,7 +33,7 @@ class VTSTOR_API cDriveEnumeratorInterface
 
 public:
-    virtual eErrorCode EnumerateDrives( Vector_Drives& AddToList, U32& Count ) = 0;
+    virtual eErrorCode EnumerateDrive( String DevicePath, Vector_Drives& AddToList, U32& Count ) = 0;
```
Use const reference to avoid creating a copy!

phandinhlan commented on vtStor/DriveManager.cpp:
```diff
@@ -30,6 +34,21 @@ cDriveManager::~cDriveManager()
+
+void cDriveManager::ResetInstance()
+{
+    delete s_Instance;
```
Where's the check for null?

phandinhlan commented on vtStor/DriveManager.cpp:
```diff
@@ -17,8 +17,12 @@ limitations under the License.
+
 namespace vtStor
 {
+    cDriveManager* cDriveManager::s_Instance = nullptr;
```
Why not use smart pointer?

phandinhlan commented on vtStor/DriveManager.cpp:
```diff
@@ -43,17 +62,23 @@ eErrorCode cDriveManager::EnumerateDrives( eScanForHardwareChanges ScanForHardwa
+        for (auto& enumerator : m_DriveEnumerators)
         {
-            //TODO: handle error
+            error = enumerator->EnumerateDrive(devicePath, m_Drives, count);
```
How does this prevent more than one enumerator from enumerating the same device?

dientc committed:
- Add public GetBusType() function in cDrive to get bus tpye.
That helps users know what kind of protocol and commandhandler need to be initialized. (+5 squashed commit)
- Using smart pointer for s_Instance in cDriveManager
 - Change the way to enumerate drives for using Range-based for loop
- Remove m_deviceHandle from Protocol class, we pass device handle as a parameter.
- CloseHandle() when remove a drive (not yet for multi-platform)

phandinhlan commented on vtStor/Drive.cpp:
```diff
@@ -41,13 +42,28 @@ void cDrive::RegisterCommandHandler(U32 CommandType, std::shared_ptr<cCommandHan
 
-    DeviceHandle& deviceHandle = commandDescriptor.GetDeviceHandle();
-    deviceHandle = m_DeviceHandle;
+eBusTypes cDrive::GetBusType()
```
Bus type would never change correct? Just get it one time and store it.

dientc replied:
Yes.

phandinhlan commented on vtStor/CommandHandlerInterface.h:
```diff
@@ -40,7 +40,7 @@ class VTSTOR_API cCommandHandlerInterface
 
 public:
-    virtual eErrorCode IssueCommand( std::shared_ptr<const cBufferInterface> CommandDescriptor, std::shared_ptr<cBufferInterface> Data ) = 0;
+    virtual eErrorCode IssueCommand( DeviceHandle Handle, std::shared_ptr<const cBufferInterface> CommandDescriptor, std::shared_ptr<cBufferInterface> Data ) = 0;
```
Use const DeviceHandle& Handle.

dientc replied:
Yes.

phandinhlan commented on vtStor/DriveManager.cpp:
```diff
@@ -30,6 +34,15 @@ cDriveManager::~cDriveManager()
+    {
+        s_Instance = std::unique_ptr<cDriveManager>(new cDriveManager());
+    }
+    DriveManager = std::move(s_Instance);
```
Reference from s_Instance would be lost? Why not just use shared_ptr? It's still a singleton with shared_ptr.

dientc replied:
Yes, I see.

dientc committed:
- Fix drive enumerating issue

dientc committed:
- Use const DeviceHandle& instead of DeviceHandle
- Let bus type as a data member

dientc committed:
- Use shared_ptr<cDriveManager> instead of unique_ptr

dientc committed:
- Use enum class rather than enum, remove plural one: eBusTypes
- Use a success flag to check EnumerateDrive()

dientc committed:
Let cDriveAta::GetBusType return eBusType::Ata directly.
